ddl/backfilling.go

```diff
@@ -520,106 +522,111 @@ func splitTableRanges(t table.PhysicalTable, store kv.Storage, startKey, endKey
+	go func() {
+		reorgInfo := scheduler.reorgInfo
+		err := consumeResults(scheduler, s, start, totalAddedCount)
+		if err != nil {
+			logutil.BgLogger().Warn("[ddl] backfill worker handle tasks failed",
+				zap.Int64("total added count", *totalAddedCount),
+				zap.String("start key", hex.EncodeToString(start)),
+				zap.String("task failed error", err.Error()))
+			s.err = err
+		} else {
+			logutil.BgLogger().Info("[ddl] backfill workers successfully processed",
+				zap.Stringer("element", reorgInfo.currElement),
+				zap.Int64("total added count", *totalAddedCount),
+				zap.String("start key", hex.EncodeToString(start)))
 		}
-		scheduler.taskCh <- task
-	}
+		s.wg.Done()
+	}()
+}
 
-	startKey := batchTasks[0].startKey
-	startTime := time.Now()
-	nextKey, taskAddedCount, err := waitTaskResults(scheduler, batchTasks, totalAddedCount)
-	elapsedTime := time.Since(startTime)
-	if err == nil {
-		err = dc.isReorgRunnable(reorgInfo.Job.ID, false)
-	}
+func (s *resultConsumer) getResult() error {
+	s.wg.Wait()
+	return s.err
+}
 
-	// Update the reorg handle that has been processed.
-	err1 := reorgInfo.UpdateReorgMeta(nextKey, scheduler.sessPool)
+func (s *resultConsumer) shouldAbort() bool {
+	return s.hasError.Load()
+}
 
-	if err != nil {
-		metrics.BatchAddIdxHistogram.WithLabelValues(metrics.LblError).Observe(elapsedTime.Seconds())
-		logutil.BgLogger().Warn("[ddl] backfill worker handle batch tasks failed",
-
-			zap.Int64("total added count", *totalAddedCount),
-			zap.String("start key", hex.EncodeToString(startKey)),
-			zap.String("next key", hex.EncodeToString(nextKey)),
-			zap.Int64("batch added count", taskAddedCount),
-			zap.String("task failed error", err.Error()),
-			zap.String("take time", elapsedTime.String()),
-			zap.NamedError("updateHandleError", err1))
+func consumeResults(scheduler *backfillScheduler, consumer *resultConsumer, start kv.Key, totalAddedCount *int64) error {
+	keeper := newDoneTaskKeeper(start)
+	handledTaskCnt := 0
+	var firstErr error
+	for result := range scheduler.resultCh {
+		err := handleOneResult(result, scheduler, consumer, keeper, totalAddedCount, handledTaskCnt)
+		handledTaskCnt++
+		if err != nil && firstErr == nil {
+			consumer.hasError.Store(true)
+			firstErr = err
+		}
+	}
+	return firstErr
+}
+
+func handleOneResult(result *backfillResult, scheduler *backfillScheduler, consumer *resultConsumer,
+	keeper *doneTaskKeeper, totalAddedCount *int64, taskSeq int) error {
+	reorgInfo := scheduler.reorgInfo
+	if result.err != nil {
+		logutil.BgLogger().Warn("[ddl] backfill worker failed",
+			zap.Int64("job ID", reorgInfo.ID),
+			zap.String("result next key", hex.EncodeToString(result.nextKey)),
+			zap.Error(result.err))
+		// Drain tasks to make it quit early.
+		for len(scheduler.taskCh) > 0 {
+			<-scheduler.taskCh
+		}
+		return result.err
+	}
+	*totalAddedCount += int64(result.addedCount)
+	keeper.updateNextKey(result.taskID, result.nextKey)
+	if taskSeq%scheduler.workerSize()*4 == 0 {
```

Comment:
what is "taskSeq%scheduler.workerSize()*4" means? also here the 4 meaning?

Reply:
For `txn` and `txn-merge` mode, it gives a chance to adjust the worker size according to `tidb_ddl_reorg_worker_cnt` for every 4 tasks processed.

For the `ingest` mode, it doesn't affect anything, because when the second task returns, all tasks have been completed.

`4` is a magic number. Any suggestion?

Reply:
should it be like this taskSeq%(scheduler.workerSize()*4), it is 4 times worker cnt?

Reply:
Updated.